File: kubilitics-backend/internal/models/metrics.go
// Package models defines the canonical metrics domain model for Kubilitics.
// All metrics queries are derived from ResourceIdentity; no resource-specific branching.

package models

// ResourceType is the canonical Kubernetes resource kind (lowercase, singular).
// Used to resolve metrics source: pod/node are direct; controllers aggregate from owned pods.
type ResourceType string

const (
	ResourceTypePod          ResourceType = "pod"
	ResourceTypeNode         ResourceType = "node"
	ResourceTypeDeployment   ResourceType = "deployment"
	ResourceTypeReplicaSet   ResourceType = "replicaset"
	ResourceTypeStatefulSet  ResourceType = "statefulset"
	ResourceTypeDaemonSet    ResourceType = "daemonset"
	ResourceTypeJob          ResourceType = "job"
	ResourceTypeCronJob      ResourceType = "cronjob"
	ResourceTypeService      ResourceType = "service"
	ResourceTypeNamespace   ResourceType = "namespace"
)

// OwnerRef identifies a controlling resource (e.g. ReplicaSet owned by Deployment).
// Used to avoid double-counting and to resolve controller → pods.
type OwnerRef struct {
	APIVersion string `json:"apiVersion"`
	Kind       string `json:"kind"`
	Name       string `json:"name"`
	UID        string `json:"uid,omitempty"`
}

// ResourceIdentity is the canonical identity for any Kubernetes resource in metrics.
// All metrics APIs accept this; resolution (controller → pods, etc.) is internal.
type ResourceIdentity struct {
	ClusterID    string       `json:"cluster_id"`
	Namespace    string       `json:"namespace"`     // empty for cluster-scoped (e.g. node)
	ResourceType ResourceType `json:"resource_type"`
	ResourceName string       `json:"resource_name"`
	OwnerRefs    []OwnerRef   `json:"owner_refs,omitempty"`
}

// Valid returns true if identity has required fields for a metrics query.
func (r *ResourceIdentity) Valid() bool {
	if r.ClusterID == "" || r.ResourceName == "" || r.ResourceType == "" {
		return false
	}
	// namespace required for namespaced types
	namespaced := map[ResourceType]bool{
		ResourceTypePod: true, ResourceTypeDeployment: true, ResourceTypeReplicaSet: true,
		ResourceTypeStatefulSet: true, ResourceTypeDaemonSet: true, ResourceTypeJob: true,
		ResourceTypeCronJob: true, ResourceTypeService: true, ResourceTypeNamespace: true,
	}
	if namespaced[r.ResourceType] && r.Namespace == "" {
		return false
	}
	return true
}

// IsController returns true if this resource type aggregates metrics from owned pods.
func (r ResourceType) IsController() bool {
	switch r {
	case ResourceTypeDeployment, ResourceTypeReplicaSet, ResourceTypeStatefulSet,
		ResourceTypeDaemonSet, ResourceTypeJob, ResourceTypeCronJob, ResourceTypeService:
		return true
	default:
		return false
	}
}

// ContainerUsage is per-container CPU/memory from Metrics Server.
type ContainerUsage struct {
	Name   string `json:"name"`
	CPU    string `json:"cpu"`    // e.g. "10.50m"
	Memory string `json:"memory"` // e.g. "32.00Mi"
}

// PodUsage is usage for a single pod (from Metrics Server or aggregation).
type PodUsage struct {
	Name           string            `json:"name"`
	Namespace      string            `json:"namespace"`
	CPU            string            `json:"cpu"`
	Memory         string            `json:"memory"`
	Containers     []ContainerUsage  `json:"containers,omitempty"`
	NetworkRxBytes int64             `json:"network_rx_bytes,omitempty"` // from kubelet stats/summary
	NetworkTxBytes int64             `json:"network_tx_bytes,omitempty"`
}

// MetricsSummary is the unified response for GET /metrics/summary.
// Same shape for pod (single), controller (aggregated), or node.
type MetricsSummary struct {
	// Identity echoed for correlation
	ClusterID    string       `json:"cluster_id"`
	Namespace    string       `json:"namespace"`
	ResourceType ResourceType `json:"resource_type"`
	ResourceName string       `json:"resource_name"`
	// Usage
	TotalCPU       string      `json:"total_cpu"`    // e.g. "125.00m"
	TotalMemory    string      `json:"total_memory"` // e.g. "256.50Mi"
	TotalNetworkRx int64       `json:"total_network_rx,omitempty"` // bytes received (all pods)
	TotalNetworkTx int64       `json:"total_network_tx,omitempty"` // bytes transmitted (all pods)
	PodCount       int         `json:"pod_count"`   // 1 for pod, N for controller
	Pods           []PodUsage  `json:"pods,omitempty"` // per-pod breakdown for controllers
	// Observability: why data might be missing (no silent failures)
	Source   string `json:"source"`   // e.g. "metrics_server"
	Warning  string `json:"warning,omitempty"`  // e.g. "3 pods skipped (not yet scheduled)"
}

// MetricsQueryResult wraps summary with query metadata for observability.
type MetricsQueryResult struct {
	Summary   *MetricsSummary `json:"summary"`
	QueryMs   int64           `json:"query_ms,omitempty"`
	CacheHit  bool            `json:"cache_hit,omitempty"`
	Error     string          `json:"error,omitempty"`
	ErrorCode string          `json:"error_code,omitempty"` // e.g. "CLUSTER_NOT_FOUND", "METRICS_SERVER_UNAVAILABLE"
}

// ---------------------------------------------------------------------------
// Historical metrics (ring-buffer backed, in-memory)
// ---------------------------------------------------------------------------

type ContainerHistoryPoint struct {
	Name      string  `json:"name"`
	CPUMilli  float64 `json:"cpu_milli"`
	MemoryMiB float64 `json:"memory_mib"`
}

type PodHistoryPoint struct {
	Name       string                  `json:"name"`
	CPUMilli   float64                 `json:"cpu_milli"`
	MemoryMiB  float64                 `json:"memory_mib"`
	NetworkRx  int64                   `json:"network_rx,omitempty"`
	NetworkTx  int64                   `json:"network_tx,omitempty"`
	Containers []ContainerHistoryPoint `json:"containers,omitempty"`
}

type MetricsHistoryPoint struct {
	Timestamp  int64              `json:"ts"`
	CPUMilli   float64            `json:"cpu_milli"`
	MemoryMiB  float64            `json:"memory_mib"`
	NetworkRx  int64              `json:"network_rx,omitempty"`
	NetworkTx  int64              `json:"network_tx,omitempty"`
	PodPoints  []PodHistoryPoint  `json:"pods,omitempty"`
}

type MetricsHistoryResponse struct {
	ClusterID    string               `json:"cluster_id"`
	Namespace    string               `json:"namespace"`
	ResourceType ResourceType         `json:"resource_type"`
	ResourceName string               `json:"resource_name"`
	Points       []MetricsHistoryPoint `json:"points"`
	IntervalSec  int                  `json:"interval_sec"`
	MaxDuration  string               `json:"max_duration"`
}

File: kubilitics-frontend/src/services/api/metrics.ts
/**
 * Metrics endpoints (pod, node, deployment, workload, summary, history).
 */
import { backendRequest } from './client';
import type {
  BackendPodMetrics,
  BackendNodeMetrics,
  BackendDeploymentMetrics,
  BackendMetricsQueryResult,
  MetricsHistoryResponse,
} from './types';

/**
 * GET /api/v1/clusters/{clusterId}/metrics/{namespace}/{pod} — pod CPU/Memory from Metrics Server.
 */
export async function getPodMetrics(
  baseUrl: string,
  clusterId: string,
  namespace: string,
  podName: string
): Promise<BackendPodMetrics> {
  const path = `clusters/${encodeURIComponent(clusterId)}/metrics/${encodeURIComponent(namespace)}/${encodeURIComponent(podName)}`;
  return backendRequest<BackendPodMetrics>(baseUrl, path);
}

/**
 * GET /api/v1/clusters/{clusterId}/metrics/nodes/{nodeName} — node CPU/Memory from Metrics Server.
 */
export async function getNodeMetrics(
  baseUrl: string,
  clusterId: string,
  nodeName: string
): Promise<BackendNodeMetrics> {
  const path = `clusters/${encodeURIComponent(clusterId)}/metrics/nodes/${encodeURIComponent(nodeName)}`;
  return backendRequest<BackendNodeMetrics>(baseUrl, path);
}

/**
 * GET /api/v1/clusters/{clusterId}/metrics/{namespace}/deployment/{name} — deployment aggregated pod metrics.
 */
export async function getDeploymentMetrics(
  baseUrl: string,
  clusterId: string,
  namespace: string,
  deploymentName: string
): Promise<BackendDeploymentMetrics> {
  const path = `clusters/${encodeURIComponent(clusterId)}/metrics/${encodeURIComponent(namespace)}/deployment/${encodeURIComponent(deploymentName)}`;
  return backendRequest<BackendDeploymentMetrics>(baseUrl, path);
}

/** Workload metrics use the same shape as deployment (aggregated + per-pod). */
export async function getReplicaSetMetrics(
  baseUrl: string,
  clusterId: string,
  namespace: string,
  name: string
): Promise<BackendDeploymentMetrics> {
  const path = `clusters/${encodeURIComponent(clusterId)}/metrics/${encodeURIComponent(namespace)}/replicaset/${encodeURIComponent(name)}`;
  return backendRequest<BackendDeploymentMetrics>(baseUrl, path);
}

export async function getStatefulSetMetrics(
  baseUrl: string,
  clusterId: string,
  namespace: string,
  name: string
): Promise<BackendDeploymentMetrics> {
  const path = `clusters/${encodeURIComponent(clusterId)}/metrics/${encodeURIComponent(namespace)}/statefulset/${encodeURIComponent(name)}`;
  return backendRequest<BackendDeploymentMetrics>(baseUrl, path);
}

export async function getDaemonSetMetrics(
  baseUrl: string,
  clusterId: string,
  namespace: string,
  name: string
): Promise<BackendDeploymentMetrics> {
  const path = `clusters/${encodeURIComponent(clusterId)}/metrics/${encodeURIComponent(namespace)}/daemonset/${encodeURIComponent(name)}`;
  return backendRequest<BackendDeploymentMetrics>(baseUrl, path);
}

export async function getJobMetrics(
  baseUrl: string,
  clusterId: string,
  namespace: string,
  name: string
): Promise<BackendDeploymentMetrics> {
  const path = `clusters/${encodeURIComponent(clusterId)}/metrics/${encodeURIComponent(namespace)}/job/${encodeURIComponent(name)}`;
  return backendRequest<BackendDeploymentMetrics>(baseUrl, path);
}

export async function getCronJobMetrics(
  baseUrl: string,
  clusterId: string,
  namespace: string,
  name: string
): Promise<BackendDeploymentMetrics> {
  const path = `clusters/${encodeURIComponent(clusterId)}/metrics/${encodeURIComponent(namespace)}/cronjob/${encodeURIComponent(name)}`;
  return backendRequest<BackendDeploymentMetrics>(baseUrl, path);
}

/**
 * GET /api/v1/clusters/{clusterId}/metrics/summary?namespace=&resource_type=&resource_name=
 * Unified, resource-agnostic metrics. Use for all resource types (pod, node, deployment, replicaset, etc.).
 */
export async function getMetricsSummary(
  baseUrl: string,
  clusterId: string,
  params: { namespace?: string; resource_type: string; resource_name: string }
): Promise<BackendMetricsQueryResult> {
  const search = new URLSearchParams();
  if (params.namespace != null && params.namespace !== '') search.set('namespace', params.namespace);
  search.set('resource_type', params.resource_type);
  search.set('resource_name', params.resource_name);
  const path = `clusters/${encodeURIComponent(clusterId)}/metrics/summary?${search.toString()}`;
  return backendRequest<BackendMetricsQueryResult>(baseUrl, path);
}

// --- Metrics History ---

export async function getMetricsHistory(
  baseUrl: string,
  clusterId: string,
  params: { namespace?: string; resource_type: string; resource_name: string; duration?: string }
): Promise<MetricsHistoryResponse> {
  const search = new URLSearchParams();
  if (params.namespace) search.set('namespace', params.namespace);
  search.set('resource_type', params.resource_type);
  search.set('resource_name', params.resource_name);
  if (params.duration) search.set('duration', params.duration);
  const path = `clusters/${encodeURIComponent(clusterId)}/metrics/history?${search.toString()}`;
  return backendRequest<MetricsHistoryResponse>(baseUrl, path);
}
